src/api/mockCourseApi.js

```diff
@@ -0,0 +1,107 @@
+import delay from './delay';
+
+// This file mocks a web API by working with the hard-coded data below.
+// It uses setTimeout to simulate the delay of an AJAX call.
+// All calls return promises.
+const courses = [
+  {
+    id: "react-flux-building-applications",
+    title: "Building Applications in React and Flux",
+    watchHref: "http://www.pluralsight.com/courses/react-flux-building-applications",
+    authorId: "cory-house",
+    length: "5:08",
+    category: "JavaScript"
+  },
+  {
+    id: "clean-code",
+    title: "Clean Code: Writing Code for Humans",
+    watchHref: "http://www.pluralsight.com/courses/writing-clean-code-humans",
+    authorId: "cory-house",
+    length: "3:10",
+    category: "Software Practices"
+  },
+  {
+    id: "architecture",
+    title: "Architecting Applications for the Real World",
+    watchHref: "http://www.pluralsight.com/courses/architecting-applications-dotnet",
+    authorId: "cory-house",
+    length: "2:52",
+    category: "Software Architecture"
+  },
+  {
+    id: "career-reboot-for-developer-mind",
+    title: "Becoming an Outlier: Reprogramming the Developer Mind",
+    watchHref: "http://www.pluralsight.com/courses/career-reboot-for-developer-mind",
+    authorId: "cory-house",
+    length: "2:30",
+    category: "Career"
+  },
+  {
+    id: "web-components-shadow-dom",
+    title: "Web Component Fundamentals",
+    watchHref: "http://www.pluralsight.com/courses/web-components-shadow-dom",
+    authorId: "cory-house",
+    length: "5:10",
+    category: "HTML5"
+  }
+];
+
+function replaceAll(str, find, replace) {
+  return str.replace(new RegExp(find, 'g'), replace);
+}
+
+//This would be performed on the server in a real app. Just stubbing in.
+const generateId = (course) => {
+  return replaceAll(course.title, ' ', '-');
+};
+
+class CourseApi {
+  static getAllCourses() {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        resolve(Object.assign([], courses));
+      }, delay);
+    });
+  }
+
+  static saveCourse(course) {
+    course = Object.assign({}, course); // to avoid manipulating object passed in.
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        // Simulate server-side validation
+        const minCourseTitleLength = 1;
+        if (course.title.length < minCourseTitleLength) {
+          reject(`Title must be at least ${minCourseTitleLength} characters.`);
+        }
+
+        if (course.id) {
+          const existingCourseIndex = courses.findIndex(a => a.id == course.id);
+          courses.splice(existingCourseIndex, 1, course);
```

Comment:
I guess we should scrap the whole idea of using splice to remove a value from an array.
It comes in handy in some scenarios, not just this one. Since we know the id of the value we could easily just filter that out and return a new array without mutating the existing one.

make courses let 

```
let courses = []

cources =  courses.filter(a => a.id !== course.id);
```